tango/integrations/gs/common.py

```diff
@@ -0,0 +1,230 @@
+                committed = False
+
+        assert name is not None, "Folder is not a GCSDataset, should not have happened."
+        return GCSDataset(name, dataset_path, created, committed)
+
+    @classmethod
+    def from_env(cls, bucket_name: str):
+        return cls(bucket_name, token="google_default")
+
+    @property
+    def full_name(self):
+        return self.bucket_name
+
+    @classmethod
+    def dataset_url(cls, workspace_url: str, dataset_name: str) -> str:
+        return os.path.join(workspace_url, dataset_name)
+
+    def get(self, dataset: Union[str, GCSDataset]) -> GCSDataset:
+        if isinstance(dataset, str):
+            path = os.path.join(self.bucket_name, dataset)
+        else:
+            # We have a dataset, and we recreate it with refreshed info.
+            path = dataset.dataset_path
+        try:
+            return self._convert_ls_info_to_dataset(self.gcs_fs.ls(path=path, detail=True))
+        except FileNotFoundError:
+            raise RemoteDatasetNotFound()
+
+    def create(self, dataset: str, commit: bool = False):
+        # since empty folders cannot exist by themselves.
+        folder_path = os.path.join(self.bucket_name, dataset)
+        try:
+            info = self.gcs_fs.info(folder_path)
+            if info["type"] == "directory":
+                raise RemoteDatasetConflict(f"{folder_path} already exists!")
+            else:
+                # Hack. Technically, this means that a folder of the name doesn't exist.
+                # A file may still exist. Ideally, shouldn't happen.
+                raise FileNotFoundError
+        except FileNotFoundError:
+            self.gcs_fs.touch(os.path.join(folder_path, self.placeholder_file), truncate=False)
+            if not commit:
+                self.gcs_fs.touch(os.path.join(folder_path, self.uncommitted_file), truncate=False)
+
+        return self._convert_ls_info_to_dataset(self.gcs_fs.ls(folder_path, detail=True))
+
+    def delete(self, dataset: GCSDataset):
+        self.gcs_fs.rm(dataset.dataset_path, recursive=True)
+
+    def sync(self, dataset: Union[str, GCSDataset], objects_dir: Path):
+        if isinstance(dataset, str):
+            folder_path = os.path.join(self.bucket_name, dataset)
+        else:
+            folder_path = dataset.dataset_path
+        # Does not remove files that may have been present before, but aren't now.
+        # TODO: potentially consider gsutil rsync with --delete --ignore-existing command
+        # Using gsutil programmatically:
+        # https://github.com/GoogleCloudPlatform/gsutil/blob/84aa9af730fe3fa1307acc1ab95aec684d127152/gslib/tests/test_rsync.py
+        try:
+            source = str(objects_dir)
+            if objects_dir.is_dir():
+                source += "/*"
+            for path in glob.glob(source):
+                basepath = os.path.basename(path)
+                self.gcs_fs.put(path, os.path.join(folder_path, basepath), recursive=True)
```

Comment:
The trouble is that that google buckets cannot handle the concept of an empty folder -- you can't create it programmatically. The file structure is flat. That's why I use a placeholder file.

Reply:
This is the ridiculousness that I'm trying to debug right now:

The following code in Python Console:

```
import gcsfs
source = 'abc'  # this is a folder
target = 'allennlp-tango-bucket/folder'
fs = gcsfs.GCSFileSystem(consistency="md5")
fs.touch(target + "/.placeholder", truncate=True)
fs.put(source, target, recursive=True)
```

works perfectly - the files and folders in `abc` are present at `allennlp-tango-bucket/folder/file.txt`, `allennlp-tango-bucket/folder/subfolder` etc.

Put the EXACT same code in a python file and run that.

The folder `abc` is copied to the target, so that I end up with `allennlp-tango-bucket/abc`.

Reply:
I've opened an issue here: https://github.com/fsspec/gcsfs/issues/517. 
In the meantime, I've changed to a slightly cleaner workaround.